### PR TITLE
readline: added RecallWordLast

### PR DESCRIPTION
**Pull request by tiymat**
Added an Action to recall the last word of the previous line, regardless of its numerical position.

### CODE REVIEW

**Comment by tiymat** [comment]
Hmm, looking at it now, I called it RecallWordLast to be consistent with the other RecallWord actions, but maybe RecallLastWord would be preferable?

File: utils/readline/hotkey_functions.go [hunk 1]
```diff
@@ -139,35 +139,58 @@ func HkFnCancelAction(rl *Instance) {
...
-	if i > len(tokens) {
-		rl.ForceHintTextUpdate(fmt.Sprintf("%s %d: previous line contained fewer words", errCannotRecallWord, i))
+	printRecalledWord(tokens[i-1], rl)
+}
+
+func hkFnRecallWordLast(rl *Instance) {
+	tokens := getLastLineTokens(rl)
+	if tokens == nil {
 		return
+	} else if len(tokens) == 0 {
```
Comment (ccoVeille) [inline utils/readline/hotkey_functions.go:174]:
```suggestion
	}

if len(tokens) == 0 {
```

No need for an `else` here

File: utils/readline/hotkey_functions.go [hunk 2]
```diff
@@ -139,35 +139,58 @@ func HkFnCancelAction(rl *Instance) {
...
 const errCannotRecallWord = "Cannot recall word"
 
 func hkFnRecallWord(rl *Instance, i int) {
-	line, err := rl.History.GetLine(rl.History.Len() - 1)
-	if err != nil {
-		rl.ForceHintTextUpdate(fmt.Sprintf("%s %d: empty history", errCannotRecallWord, i))
+	tokens := getLastLineTokens(rl)
+	if tokens == nil {
+		return
+	} else if i > len(tokens) {
```
Comment (ccoVeille) [inline utils/readline/hotkey_functions.go:162]: No need for an else here


```suggestion
	}

if i > len(tokens) {
```

File: utils/readline/hotkey_functions.go [hunk 3]
```diff
@@ -139,35 +139,58 @@ func HkFnCancelAction(rl *Instance) {
...
 		return
+	} else if len(tokens) == 0 {
+		rl.ForceHintTextUpdate(fmt.Sprintf("%s: previous line contained no words", errCannotRecallWord))
 	}
 
-	word := rTrimWhiteSpace(tokens[i-1])
+	printRecalledWord(tokens[len(tokens)-1], rl)
+}
+
+func getLastLineTokens(rl *Instance) []string {
```
Comment (ccoVeille) [inline utils/readline/hotkey_functions.go:181]: I would have expected to see
```suggestion
func getLastLineTokens(rl *Instance) []string, error {
```

File: utils/readline/hotkey_functions.go [hunk 4]
```diff
@@ -139,35 +139,58 @@ func HkFnCancelAction(rl *Instance) {
...
+	line, err := rl.History.GetLine(rl.History.Len() - 1)
+	if err != nil {
+		rl.ForceHintTextUpdate(fmt.Sprintf("%s: empty history", errCannotRecallWord))
```
Comment (ccoVeille) [inline utils/readline/hotkey_functions.go:184]: How can you be sure the error is not a major error to handle differently

File: utils/readline/hotkey_functions.go (hunk 1 above)
Comment (lmorg) [reply utils/readline/hotkey_functions.go:174]: In fairness, a slice can be 0 length but also not nil. eg https://go.dev/play/p/Cnfpc6s6l2e

Though changing a nil check for a `len` check would solve that problem

**Comment by lmorg** [comment]
@tiymat I've created an alternative PR, basically borrowing the same ideas as yourself but with some different thoughts on error handling.

I don't mind if you want to incorporate these ideas into your own, or just test this code for merging.

https://github.com/lmorg/murex/pull/810

**Comment by tiymat** [comment]
Yes, the error-handling is much more consistent in your implementation than mine. Thanks for adding it!

And thanks for the review @ccoVeille !

**Comment by lmorg** [comment]
@tiymat i stole your modularisation  and naming conventions though. Good work and thanks for the PR.

File: utils/readline/hotkey_functions.go (hunk 1 above)
Comment (ccoVeille) [reply utils/readline/hotkey_functions.go:174]: Yes, indeed

**Comment by ccoVeille** [comment]
You're welcome. I'll have a look at the new PR

**Pull request closed without merging**